File: Week 6/sentimental-credit/credit.py
# This program verify if a credit card is an American Express, MasterCard, Visa or none of them
from cs50 import get_int


def main():

    # promt the user for the credit card number
    number = credit_number()
    # print the type of the credit card
    check_type(number)


def credit_number():
    n = -1
    while n < 0:
        n = get_int("Number: ")
    return n


# checks if the checksum is valid
def checksum(n):
    return (Luhn_sum(n) % 10) == 0


# checks if the credit card is valid and print his type
def check_type(n):
    type = "INVALID"
    f2d = first2(n)
    f1d = round(f2d/10)

    if checksum(n):
        if digits(n) == 15 and (f2d == 34 or f2d == 37):
            type = "AMEX"
        elif (digits(n) == 13 or digits(n) == 16) and f1d == 4:
            type = "VISA"
        elif digits(n) == 16 and (f2d == 51 or f2d == 52 or f2d == 53 or f2d == 54 or f2d == 55):
            type = "MASTERCARD"

    print(f"{type}")


# return the digits of a number
def digits(n):
    n = str(n)
    return len(n)


# calculate the Luhn's sum
def Luhn_sum(n):
    # convert the credit card number to a list of digits
    def digits_of(e):
        return [int(d) for d in str(e)]
    digits = digits_of(n)

    # generate a list of odd and even digits starting from the rightmost or last digit
    odd_digits = digits[-1::-2]
    even_digits = digits[-2::-2]

    # multiply each value from the even_digits by 2. If any number is greater than 9, sum its digits to convert it into a single-digit number
    checksum = 0
    for d in even_digits:
        checksum += sum(digits_of(d*2))

    # sum all the odd numbers along with this checksum
    checksum += sum(odd_digits)

    return checksum


# return the first two digits of a number
def first2(n):
    n = str(n)
    fd = n[:2]
    return int(fd)


main()

File: Week 1/credit/credit.c
//This program verify if a credit card is an American Express, MasterCard, Visa or none of them

#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

// Functions
long credit_number(void);
bool checksum(long n);
void check_type(long n);
int digits(long n);
int Luhn_sum(long n);
int first_2(long n);

int main(void)
{
    bool check1;
    // Asks for the credit card number
    long number = credit_number();
    // Print the type of the credit card
    check_type(number);

}

//Checks if the credit card is valid and print his type
void check_type(long n)
{
    char type[20] = "INVALID";
    int fd = first_2(n);
    if (checksum(n))
    {
        if (digits(n) == 15 && (fd == 34 || fd == 37))
        {
            strcpy(type, "AMEX");
        }
            else if ((digits(n) == 13 || digits(n) == 16) && (fd/10 == 4))
            {
                strcpy(type, "VISA");
            }
                else if (digits(n) == 16 && (fd == 51 || fd == 52 || fd == 53 || fd == 54 || fd == 55))
                {
                    strcpy(type, "MASTERCARD");
                }
    }

    printf("%s\n", type);
}

// Calculate the first two digits of a credit card
int first_2(long n)
{
    return (n / pow(10, (digits(n) - 2) ));
}

long credit_number()
{
    long n;
    do
    {
        n = get_long("Number: ");

    }
    while (n < 0);
    return n;
}

// Checks if the checksum is valid
bool checksum(long n)
{
    int mod = Luhn_sum(n) % 10;
    return mod == 0;
}

int Luhn_sum(long n)
{
    long aux0 = n;
    int aux;
    int sum = 0;

    // Adds the first part of Luhn's algorithm to the sum
    do
    {
        n = n / 10;
        aux = (n % 10) * 2;
        if (aux > 9)
        {
            aux = (aux / 10) + (aux % 10);
        }
        sum = sum + aux;
        n = n / 10;
    }
    while (n > 0);

    // Adds the second part of Luhn's algorithm to the sum
    do
    {
        sum = sum + (aux0 % 10);
        aux0 = aux0 / 100;
    }
    while (aux0 > 0);
    return sum;
}

// Counts the digits of a number
int digits(long n)
{
    int count = 0;;
    do
    {
        n /= 10;
        ++count;
    }
    while (n != 0);

    return count;
}